{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hierarchical clustering of Pa 14 resistant samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook we are going to analyse the merged output files of SNPeff variants annotation and SnpSift fields extract to build a hierarchical clustering tree of the samples."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This line gives us a larger view of the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we install the necessary library.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching package metadata ...........\n",
      "Solving package specifications: ..........\n",
      "\n",
      "# All requested packages already installed.\n",
      "# packages in environment at /opt/conda/envs/python2:\n",
      "#\n",
      "biopython                 1.69                np112py27_0  \n",
      "Fetching package metadata .............\n",
      "Solving package specifications: ..........\n",
      "Warning: 2 possible package resolutions (only showing differing packages):\n",
      "  - anaconda::intel-openmp-2018.0.0-8.tar.bz2\n",
      "  - anaconda::intel-openmp-2018.0.0-hc7b2577_8.tar.bz2\n",
      "\n",
      "Package plan for installation in environment /opt/conda/envs/python2:\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    blas-1.0                   |              mkl           6 KB  anaconda\n",
      "    intel-openmp-2018.0.0      |                8         620 KB  anaconda\n",
      "    libgcc-ng-7.2.0            |       hdf63c60_3         6.1 MB  anaconda\n",
      "    libgfortran-ng-7.2.0       |       hdf63c60_3         1.2 MB  anaconda\n",
      "    libstdcxx-ng-7.2.0         |       hdf63c60_3         2.5 MB  anaconda\n",
      "    mkl-2018.0.2               |                1       205.2 MB  anaconda\n",
      "    backports-1.0              |   py27h63c9359_1           3 KB  anaconda\n",
      "    bz2file-0.98               |           py27_0          10 KB  anaconda\n",
      "    contextlib2-0.4.0          |           py27_0           6 KB  anaconda\n",
      "    future-0.15.2              |           py27_0         616 KB  anaconda\n",
      "    lockfile-0.12.2            |           py27_0          14 KB  anaconda\n",
      "    natsort-4.0.3              |           py27_0          39 KB  anaconda\n",
      "    numpy-1.11.3               |   py27h3dfced4_4         3.5 MB  anaconda\n",
      "    numexpr-2.6.5              |   py27h7bf3b9c_0         187 KB  anaconda\n",
      "    qt-5.6.2                   |                3        44.7 MB  conda-forge\n",
      "    scipy-1.1.0                |   py27hfc37229_0        17.9 MB  anaconda\n",
      "    backports.functools_lru_cache-1.5|           py27_1           9 KB  anaconda\n",
      "    nose-1.3.7                 |   py27heec2199_2         212 KB  anaconda\n",
      "    pandas-0.23.0              |   py27h637b7d7_0        11.8 MB  anaconda\n",
      "    pyqt-5.6.0                 |           py27_2         5.3 MB  anaconda\n",
      "    scikit-learn-0.19.1        |   py27h445a80a_0         5.3 MB  anaconda\n",
      "    matplotlib-2.1.1           |           py27_0         9.2 MB  conda-forge\n",
      "    cachecontrol-0.11.5        |           py27_0          19 KB  anaconda\n",
      "    scikit-bio-0.4.0           |      np111py27_0         980 KB  anaconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:       315.4 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "    backports:                     1.0-py27h63c9359_1                 anaconda   \n",
      "    backports.functools_lru_cache: 1.5-py27_1                         anaconda   \n",
      "    bz2file:                       0.98-py27_0                        anaconda   \n",
      "    cachecontrol:                  0.11.5-py27_0                      anaconda   \n",
      "    contextlib2:                   0.4.0-py27_0                       anaconda   \n",
      "    future:                        0.15.2-py27_0                      anaconda   \n",
      "    intel-openmp:                  2018.0.0-8                         anaconda   \n",
      "    libgcc-ng:                     7.2.0-hdf63c60_3                   anaconda   \n",
      "    libgfortran-ng:                7.2.0-hdf63c60_3                   anaconda   \n",
      "    libstdcxx-ng:                  7.2.0-hdf63c60_3                   anaconda   \n",
      "    lockfile:                      0.12.2-py27_0                      anaconda   \n",
      "    mkl:                           2018.0.2-1                         anaconda   \n",
      "    natsort:                       4.0.3-py27_0                       anaconda   \n",
      "    nose:                          1.3.7-py27heec2199_2               anaconda   \n",
      "    pyqt:                          5.6.0-py27_2                       anaconda   \n",
      "    qt:                            5.6.2-3                            conda-forge\n",
      "    scikit-bio:                    0.4.0-np111py27_0                  anaconda   \n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "    matplotlib:                    2.0.2-np112py27_0                  conda-forge --> 2.1.1-py27_0          conda-forge\n",
      "    numexpr:                       2.6.2-np112py27_0                  conda-forge --> 2.6.5-py27h7bf3b9c_0  anaconda   \n",
      "    pandas:                        0.19.2-np112py27_1                 conda-forge --> 0.23.0-py27h637b7d7_0 anaconda   \n",
      "    scikit-learn:                  0.18.1-np112py27_blas_openblas_200 conda-forge [blas_openblas] --> 0.19.1-py27h445a80a_0 anaconda   \n",
      "    scipy:                         0.19.0-np112py27_blas_openblas_201 conda-forge [blas_openblas] --> 1.1.0-py27hfc37229_0  anaconda   \n",
      "\n",
      "The following packages will be SUPERCEDED by a higher-priority channel:\n",
      "\n",
      "    blas:                          1.1-openblas                       conda-forge --> 1.0-mkl               anaconda   \n",
      "    numpy:                         1.12.1-py27_blas_openblas_200      conda-forge [blas_openblas] --> 1.11.3-py27h3dfced4_4 anaconda   \n",
      "\n",
      "Fetching packages ...\n",
      "blas-1.0-mkl.t 100% |################################| Time: 0:00:00   3.10 MB/s\n",
      "intel-openmp-2 100% |################################| Time: 0:00:02 302.41 kB/s\n",
      "libgcc-ng-7.2. 100% |################################| Time: 0:00:31 205.85 kB/s\n",
      "libgfortran-ng 100% |################################| Time: 0:00:05 217.96 kB/s\n",
      "libstdcxx-ng-7 100% |################################| Time: 0:00:11 228.51 kB/s\n",
      "mkl-2018.0.2-1 100% |################################| Time: 0:01:34   2.27 MB/s\n",
      "backports-1.0- 100% |################################| Time: 0:00:00   2.22 MB/s\n",
      "bz2file-0.98-p 100% |################################| Time: 0:00:00   6.96 MB/s\n",
      "contextlib2-0. 100% |################################| Time: 0:00:00   4.78 MB/s\n",
      "future-0.15.2- 100% |################################| Time: 0:00:02 309.92 kB/s\n",
      "lockfile-0.12. 100% |################################| Time: 0:00:00 606.62 kB/s\n",
      "natsort-4.0.3- 100% |################################| Time: 0:00:00 542.15 kB/s\n",
      "numpy-1.11.3-p 100% |################################| Time: 0:00:17 203.70 kB/s\n",
      "numexpr-2.6.5- 100% |################################| Time: 0:00:00 297.28 kB/s\n",
      "qt-5.6.2-3.tar 100% |################################| Time: 0:00:53 884.32 kB/s\n",
      "scipy-1.1.0-py 100% |################################| Time: 0:00:34 540.09 kB/s\n",
      "backports.func 100% |################################| Time: 0:00:00   5.66 MB/s\n",
      "nose-1.3.7-py2 100% |################################| Time: 0:00:00 638.11 kB/s\n",
      "pandas-0.23.0- 100% |################################| Time: 0:00:13 903.81 kB/s\n",
      "pyqt-5.6.0-py2 100% |################################| Time: 0:00:27 201.51 kB/s\n",
      "scikit-learn-0 100% |################################| Time: 0:00:27 202.86 kB/s\n",
      "matplotlib-2.1 100% |################################| Time: 0:00:44 214.06 kB/s\n",
      "cachecontrol-0 100% |################################| Time: 0:00:00 818.64 kB/s\n",
      "scikit-bio-0.4 100% |################################| Time: 0:00:03 250.87 kB/s\n",
      "Extracting packages ...\n",
      "[      COMPLETE      ]|###################################################| 100%\n",
      "Unlinking packages ...\n",
      "[      COMPLETE      ]|###################################################| 100%\n",
      "Linking packages ...\n",
      "[      COMPLETE      ]|###################################################| 100%\n",
      "Fetching package metadata ...........\n",
      "Solving package specifications: ..........\n",
      "\n",
      "Package plan for installation in environment /opt/conda/envs/python2:\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    pandasql-0.7.3             |           py27_0          26 KB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "    pandasql: 0.7.3-py27_0\n",
      "\n",
      "Fetching packages ...\n",
      "pandasql-0.7.3 100% |################################| Time: 0:00:00   9.22 MB/s\n",
      "Extracting packages ...\n",
      "[      COMPLETE      ]|###################################################| 100%\n",
      "Linking packages ...\n",
      "[      COMPLETE      ]|###################################################| 100%\n"
     ]
    }
   ],
   "source": [
    "!conda install -y biopython\n",
    "!conda install -y -c anaconda scikit-bio \n",
    "!conda install -y pandasql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we import the useful modules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/python2/lib/python2.7/site-packages/matplotlib/font_manager.py:281: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.\n",
      "  'Matplotlib is building the font cache using fc-list. '\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pandasql import *\n",
    "from IPython.display import display\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re\n",
    "from Bio import Phylo\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator\n",
    "from Bio import AlignIO\n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor\n",
    "import pylab\n",
    "\n",
    "pysqldf = lambda q: sqldf(q, globals())\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from sklearn.metrics.pairwise import euclidean_distances\n",
    "import scipy\n",
    "from scipy.spatial.distance import squareform, pdist\n",
    "from Bio.Phylo.TreeConstruction import _Matrix\n",
    "\n",
    "from Bio.Phylo.TreeConstruction import _DistanceMatrix\n",
    "import skbio\n",
    "import os\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import SeqIO\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting Strand Bias"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The strand Bias is a sequencing bias that result in an inbalance of infered genotype from the positive and negative strand. We want to eliminate the variants where the strand bias is too high.\n",
    "\n",
    "The strand bias has been calculated on Galaxy using the tool \"compute on column\" using the following equations (from https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-13-666) :  \n",
    "\n",
    "\n",
    "Calculation Strand Bias : \n",
    "\n",
    " a -  Forward strand reference allele = SRF column (5) \n",
    "\n",
    " b - Forward strand non reference allele = SAF column (7) \n",
    "\n",
    " c - Reverse strand reference allele = SRR column (6) \n",
    "\n",
    " d - Reverse strand non reference allele = SAR column (8) \n",
    "\n",
    "``` SB: SB is defined as ∣∣(b/(a+b))−(d/(c+d))∣∣/((b+d)/(a+b+c+d))\n",
    "SB =  ∣∣(SAF/(SRF+SAF))−(SAR/(SRR+SAR))∣∣/((SAF+SAR)/(SRF+SAF+SRR+SAR))\n",
    "SB =  abs((SAF/(SRF+SAF))−(SAR/(SRR+SAR)))/((SAF+SAR)/(SRF+SAF+SRR+SAR))\n",
    "SB =  abs(((c6/(c6+c8))-(c7/(c9+c7)))) / ((c6+c7)/(c9+c6+c7+c8) ) (Formula applied in Compute Column tool)```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import of the file resulting of collection collapse snpsift output collection, and plotting of the Strand Bias repartition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA2AAAAJNCAYAAABX1xb4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4xLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvAOZPmwAAIABJREFUeJzt3XmQXmdh5/tf73u32q3WLtuyvGLH\nFgZDvEKCwSSZkGEYrjUMS2UquZkUlQozCTUEz0yooUKNmap7Ic5cTFiGuZk7Fa4xi3PBwDheMMY2\nsR3bGK+SZUtqbS2p1fve7/2jpfYqy0v301b786nq6lbrvG8/evX0Oe/3nPOet6pSqVQCAADAgqte\n7AEAAAC8UQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAopHa+77C3d3C+73JedHY2p69vZLGH\nwRJlfrGQzC8WkvnFQjPHWEiv1/nV3d121L97wxwBq62tWewhsISZXywk84uFZH6x0MwxFtLxOL/e\nMAEGAACw2AQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgA\nAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBAB\nBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhdQu9gBK+eGdT2VwaOxV3fadm9bO72AAAIA3JEfA\nAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACF\nCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAA\nQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEG\nAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChE\ngAEAABQiwAAAAAoRYAAAAIXUHmuB4eHh/Lt/9+/S39+fycnJfPzjH8+ll15aYmwAAABLyjED7Dvf\n+U42bNiQP/mTP8nevXvzsY99LD/84Q9LjA0AAGBJOeYpiJ2dnTl06FCSZGBgIJ2dnQs+KAAAgKXo\nmEfAfuu3fivf/va38+53vzsDAwP58pe/XGJcAAAAS84xA+x73/te1qxZk6997Wt59NFHc9VVV+X6\n668/6vKdnc2pra2Z10HOiy0H0tba+Kpu2t3dNs+DYSkyT1hI5hcLyfxioZljLKTjbX4dM8Duu+++\nXHLJJUmSM888M3v37s3U1FRqa1/8pn19I/M7wnk0ODT2qm7X2zs4zyNhqenubjNPWDDmFwvJ/GKh\nmWMspNfr/HqpKDzma8BOOumkPPDAA0mSnp6etLS0HDW+AAAAOLpjltSVV16ZT3/60/nwhz+cqamp\nfOYznykwLAAAgKXnmAHW0tKSL37xiyXGAgAAsKQd8xREAAAA5ocAAwAAKESAAQAAFCLAAAAAChFg\nAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBC\nBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAA\noBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAAD\nAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQi\nwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAA\nhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgA\nAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBAB\nBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAo\nRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAA\nAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgw\nAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFvKwAu+GGG/K+970v\n/+yf/bPcdtttCz0mAACAJemYAdbX15f/+l//a/7n//yfufbaa3PTTTeVGBcAAMCSU3usBe68885c\neOGFaW1tTWtraz772c+WGBcAAMCSc8wjYDt37kylUsknPvGJfOhDH8qdd95ZYlwAAABLzjGPgCXJ\n3r1781d/9VfZtWtXPvrRj+aWW25JVVXViy7b2dmc2tqaeR3kvNhyIG2tja/qpt3dbfM8GJYi84SF\nZH6xkMwvFpo5xkI63ubXMQOsq6srb37zm1NbW5sTTzwxLS0tOXjwYLq6ul50+b6+kXkf5HwZHBp7\nVbfr7R2c55Gw1HR3t5knLBjzi4VkfrHQzDEW0ut1fr1UFB7zFMRLLrkkd911V2ZmZnLw4MGMjIyk\ns7NzXgcIAADwRnDMI2ArV67MFVdckY997GMZHR3Nv//3/z7V1d4+DAAA4JV6Wa8B27x5czZv3rzQ\nYwEAAFjSHMoCAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMA\nAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLA\nAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACF\nCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAA\nQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEG\nAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChE\ngAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAA\nChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAA\nAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCEC\nDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQ\niAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEA\nABQiwAAAAAoRYAAAAIUIMAAAgEJeVoCNjY3lXe96V7797W8v9HgAAACWrJcVYF/60peybNmyhR4L\nAADAknbMANu6dWu2bNmSd77znQWGAwAAsHQdM8CuvvrqfOpTnyoxFgAAgCWt9qX+8rvf/W42bdqU\n9evXv+w77OxsTm1tzWse2LzbciBtrY2v6qbd3W3zPBiWIvOEhWR+sZDMLxaaOcZCOt7m10sG2K23\n3podO3bk1ltvzZ49e1JfX59Vq1bloosuOupt+vpG5n2Q82VwaOxV3a63d3CeR8JS093dZp6wYMwv\nFpL5xUIzx1hIr9f59VJR+JIB9oUvfGHu62uuuSZr1659yfgCAADg6LwPGAAAQCEveQTs2f7oj/5o\nIccBAACw5DkCBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEG\nAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChE\ngAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAA\nChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAA\nAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCEC\nDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQ\niAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEA\nABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFg\nAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBC\nBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAA\noBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAAD\nAAAoRIABAAAUIsAAAAAKqX05C33+85/Pvffem6mpqfzBH/xB3vOe9yz0uAAAAJacYwbYXXfdlSee\neCLf/OY309fXl/e///0CDAAA4FU4ZoBdcMEFOffcc5MkHR0dGR0dzfT0dGpqahZ8cAAAAEvJMV8D\nVlNTk+bm5iTJddddl8suu0x8AQAAvApVlUql8nIWvOmmm/LlL385X//619PW1nbU5aamplNb+/oL\ntB/e+dSrvu17Lzx5voYBAAC8gb2si3Dcfvvtufbaa/PVr371JeMrSfr6RuZlYAthcGjsVd2ut3dw\nnkfCUtPd3WaesGDMLxaS+cVCM8dYSK/X+dXdffRmOmaADQ4O5vOf/3y+8Y1vZNmyZfM6MAAAgDeS\nYwbYD37wg/T19eUTn/jE3PeuvvrqrFmzZkEHBgAAsNQcM8CuvPLKXHnllSXGAgAAsKQd8yqIAAAA\nzA8BBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAAD\nAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQi\nwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAA\nhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgA\nAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAAAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBAB\nBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgwAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAo\nRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBBgAAEAhAgwAAKAQAQYAAFCIAAMAAChEgAEAABQiwAAA\nAAoRYAAAAIUIMAAAgEIEGAAAQCECDAAAoBABBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhQgw\nAACAQgQYAABAIQIMAACgEAEGAABQiAADAAAoRIABAAAUIsAAAAAKEWAAAACFCDAAAIBCBNgrsL9/\nNKPjU4s9DAAA4DhVu9gDOB6MjE3l2z/Zmlvu68lJq9ry6Y+8JbU12hUAAHhlBNhLqFQqeXrPYL73\n023pH5pIfW11ntozmBvv3p7fvujkxR4eAABwnHEY5ygGRyby9/f25CcP7M7w6FTef+mGXP2HF2VZ\na31u+Om27Nw3tNhDBAAAjjMC7EU88nRfbvjpU9m1fziru5rz2d97W3774g3paKnPx957ZqZnKvna\n9x/J1PTMYg8VAAA4jgiw5xkencw/PLIvdbXVufTc1bn8reuysrN57u/PO3V5Lj5nVZ7eO3sqIgAA\nwMslwJ6np3c4SXLuxq5sWNOeqqqqFyyz+fLTnIoIAAC8YgLseXb2zgbV2u6Woy7T0ljnVEQAAOAV\nE2DPMj09kz0HR9LRUp+25vqXXNapiAAAwCvlMvTPsufgaKamKy84+nXr/T0vuvy6la1peqI237v9\nyUxNz6SzreEFy7xz09oFGSsAAHD8cQTsWY6cfriuu/VlLd9QV5MLz16ZmUry9/fuTN/g2EIODwAA\nOM4JsMMqlUp6eodTV1udFZ1NL/t261a05s2nL8/I2FRuvGv7XMQBAAA8nwA7rH94IkOjk1nT1Zzq\n6hde+fCl/MopXbnsvNWpVJJb7u3JI0/3LdAoAQCA45nXgB125PLza1/m6YfPd/Lq9rQ01eWW+3ry\nD4/sy8DwRC44c8Vzltl/aDSP7TiUrT39WdvdmsvOW5O6Wg38csxUKvnJ/bvypg0nZMWyl3+EEgAA\nXk8E2GEv5/Lzx9K9rCm/eeFJufnenXls+6EMjkykrrY6j28/lEe3H8qBgee+RuyHdz+d912yIRed\nsyo11ULspdz9y735v3/0WDaubc+nP/yWF31/NgAAeL0TYEkmJqezr280XR2NaWp4bQ9Ja1Nd3vur\nJ+b2B3anp3c4/+0HjyZJ6uuqc+LK1qzsbM7yZY15es9gHt1+KP/tB4/m+tuezKbTluekla0vCIv5\nuopi3+B4bvnHnTn3lOU5dV3HvNxnKVPTM/nuT59MkmztGcgDWw9k06nLF3lUAADwygmwJLsOjKRS\nSda9hqNfz1ZfW5Nfe/PabOnpz/R0Jau6mrOstf45cdW9rClnndyZB7ccyJae/vzk/l05ob0h55/e\nnTXL52ccSTI0Opkb73o6N927M5NTM/nJA7vzud9/e5ob6+btZyy02x/cnd5DYzl3Y1d+sfVAvn3b\nkzl3Y1eqHQUDAOA447y3JD37jpx++Ope//Viqqurcvr6ZTnr5M50tjW86ClzLY11ufCcVfmdSzZk\nw+q2HBwYz0337Mzf37szh4bGX9PPH5+czvfvfCqfuvbO3Hj39rQ21eWtZ3RnYHgi37p162u675Im\nJqdzwx3bUl9Xnd/9zbPyq2evys7eofzDI/sWe2gAAPCKveGPgFUqlfTsH05jfU262l/4RsoltLfU\n59Lz1uTsDWO559He9PQOZ9f+4Zy+flneesaKtDa9/KNVo+NT+dlDe/L/3flU+ocm0tJYmyt//dT8\n+vlrU1VVld3/7R9y6/27ctE5q4+LUxFvvq8n/UMT+a0LT0pHS31+59IN+fkje/Od25/MW87oTm3N\nwu5DGBiZyPW3bs3b37Qybzr5hAX9WQAALH1v+CNgB/rHMjYxnbXdLYt+YYcT2hvz7gvW5Z1vXpPW\npro8tv1QPnXtnfnxz7dnfGL6qLerVCp5fMehfO37D+ff/NVP8//8r8czOj6Vf3LRybn6X1+UK952\nYupqa1JbU52PvffMJMl//+GjmZqeOep9zsxUcscvdue2+3uydVf/S/78hTI6PpXv3/lUmhtq8963\nn5gkWbGsKZedtyb7+kbzs4f2LOjPH5uYyheveyC3P7g7X7juwTz81MEF/XmvB9MzM3lw64GMTUwt\n9lAAAJakN/wRsJ2HLz+/bh5PP3wtqqqqcuLKtqztbs1j2/vy8La+/O3NW/LNm7ekq6Mx67pbs2Z5\nS9Z2t2RlZ3Me296X2x/cnT0HR5Ikyzsac+m5q3PZprX5xyd68/NH977gZ5y+viOP7+jPl777UH5l\nY9cL/n56eiaPbj+Uex/rfWZcSbo7m7KuuzXrultyxomdOX19x4JevfFHP9+e4bGpfOAdp6TlWa9Z\n+ycXnZyf/mJ3brhjWy48e2Xqamvm/WdPTc/k//rOQ9m2ezBnbzghj23vy19+68H8m//tvJxxYue8\n/7zXg7GJqVz7vV/mwa0HcuKK1vzxB89LZ9viHBUGAFiqaj7zmc98Zj7vcGRkYj7vbt7sPjiaiRfZ\nq3/vY/syPjGdXz1n5evqUvDVVVXpXtaUj773zFRXV6WmpjqHBsfz9N6hPLGzP/c93pufPLArDz/V\nl7GJ6bztrBX5F+86LZsvPy1nntiZxvqaPLVn8EXve0VnU7b09Gf3gZGcvKotDfXPBMzE5HRuvq8n\nT+zszxnrl+W3Lz45y5c1pr62OgcGxvP03sE8tuNQfvbQntx8787D759WSVd748s+HbBSqeTex3rz\nte8/nD0HR3Lq2o4X3HZwZCLX3vDLtDTU5g/ed85z/r6poTaj41P5xZMH09ZUn41r5/dUykqlkq9/\n/9Hc93hvzt3YlT/+5+fmpFXtufvhvfn5o/ty1omdOaG98Tm3aWlpeN3O/Zejf3gi/8c3789jOw6l\nq70xew6O5N7H9uWcDV1pa65f7OEdt0bHp9KzfyjtzfWv6Qh7Q2Ndtu/qT3Nj7aIfqWfpOd7XX7z+\nmWMspNfr/GppOfpO7KpKpVI51h187nOfywMPPJCqqqp8+tOfzrnnnnvUZXt7X/xJ/2K7d8uBDA49\n9324Rsam8q1bt2bVCc15z9vWL9LIXpmxiakcGpzIoaHx9A9PpKOlPhvWtKeh7pUdBdq2eyC3P7A7\nq7uac/lb16Wqqiqj41O56Z6d6Rscz4krW3PpuatT86zwqVQqGR2fzsGBsfTsH86OvUMZGZ+N2prq\nqqzuas5Jq9ryoctPP+rl/LftHsg3//6JPL6zf+57Xe2N+cgVZ+TcjV259f6eJMk9j+7Lw0/15YKz\nVuSsk154xGlsYjrfue3J1NRU5f2XnTL3htZHLts/M1PJtj0DaWmsy4plTamufvlPWq+7ZUtuvHt7\nTlnTnk9ufvNcoN7z6L5c+71fpqG+Jp/8F5ty8qr2udt0d7e9buf+sew5OJL/8/+9P72HxnLJr6zO\nR997Rm68e3u+85Mn09JYmz/6wLk5ff2yF73t3r6RPPTkwZy2riMnrmw75s/5uzueyn1P9Oack0/I\nb1988jFv82J2HxjOA1sO5Mld/VmzfPZo7MY17al/hb8DC6FSqWTPwZE8uPVAHtiyP0/s7M/0TCVr\nu1vygcs25rxTu15RQI1PTuf2B3blf927M719o9mwuj3//B2n5CyvR+Q1mpyayT2P7suBgbH8xiWn\npGbm6KekL2UzM5U8tO1AhsemsunU5a/5rWh4ccfzNpLXv9fr/OruPvpznGMG2M9//vN87Wtfy5e/\n/OVs2bIlf/Znf5brrrvuqMu/Hh+AH969PVt29eeEtoas6GyaO5qyZWd/fvbQnrz1jO68acMb6wlN\npVLJzff2pGf/cC45d3WWdzTmpnt2Zmh0Mqev78jb3rTymJd5r1QqOTAwnh17B7Nj31AODc3ufait\nqc55G7vytjetzHkbu1JfV5ODA2O5/rYnc+cvZ1+39ebTluefXnpKfv7I3vzw7u2Znqnk7W9amRNX\ntmamUsl3frItTfU1+aeXbTjqkckHtuzPA1sOZNNpy3Pu4VMpT1vbkZ89tCd3/nLP3Hjq66qzrrs1\n61c887Guu/UFG9pb7+/Jw08dzD2P9qa9pT7vffv6NNY/d5ltuwby0wd3p66uOle8bX062xozPVPJ\nm89cmXse2p0tPf15eu9gqlKVxoaaNNXXpqmhNo31NWmsr0lNdXWqqmZPNZ39PHu0s625Psta67Os\ntSEdhz+/2icCRyLgiZ39eWLHoWzfN5R13a3ZdNrynH3yCWlufOZ+t/b054vfejBDo5N538Un53cu\n2TAXCHf8Yne+ceOjqaqqyv/+22/KW89ckWT29Mz7n9ifW/6xJ4883Td3X6esac87Nq3J285a+Zwd\nArsPDOfvfvZU7n54byqVpKWxNsNjU3Pz4H0Xb8hJq46+kpqemckTO/pz/5b9eWDL/uztG33BMrU1\nVTlldfvsqbEnLsuqzuYsa6t/WUe1K5XKqz6qNDo+ld0HRrL7wHCe2jOYX2w9kH2HnhnfSavasryj\nMfc93ptKJTl1XUf++Ts2HjVojxganczN9+2c+52sr6vJhlVteWzHoSTJm07uzAfesTEbVre/5P0s\nlEqlksmpmZcVvdMzM+npHc6BgbGs627N8o7GYz7eo+NT6R+eyPKOV3ZkveTRweGxyWztGcihofGc\nvKot67pbX9GOnvlWqVSy79BohkYns3Z5ywvWXUcc6B/Lrff35CcP7MrgyGSS2fXQr5zSlV8/f13O\nOeWE4/YtPmYqlfT0Duex7X0ZnZjO6es6csqa9hc9Tf3gwFhuf3B3fvLArvQNzl51uKG+Jhe+aWXe\nsWnti66TDg6M5cEnD+QXWw9kbGI652w4Iedu7Mqa5Yv/GvLXu66u1jyxbX9qqqvS3vLazgg4ng2M\nTOSp3YOpranK+hWtzjCZJ0sywL74xS9mzZo1+eAHP5gkueKKK3L99dentfXFXzP1enwArvrKXdl9\nYPY1UtXVVVnR2ZTVXc3ZvX8kew6O5Hcu2ZCO1jfeL8HgyERu+OlTqa2ZjYKxiemcu7HrFe+lP2Jg\neCJP7RnMtt0D6Z+LsaqsOqE5uw+MZHqmkhPaG/LWM1ZkVVfz3O36Bsdy50N7s79/LPV11VnW2pB9\nfaO56JxVL3mlxsmpmXz7ticzU6nkvI1d2bZ7IAcGZjekzQ21ecsZ3ZmarmTHvqHsPjCc6ZnnTvUV\ny5qeE2X/8Ni+3PXLvWlqqMlvvP2ktDa/+NUnt/b0545f7ElDXU3aW+pzYGAsM8+674a6mlRXV2Vy\najpT08c8wHxU9XXVqamuzkx+/1glAAAP1ElEQVSlkkqlkkolc59bmurS0VKf9pb6tDfXp6O1Pk31\nNdm+dyhP7DyUgcNPrJLZo5NH/u01h98e4bxTl6e5oTb//YePZqZSya++aWVOe5Eo2LV/OLf+Y0+m\npit582nLMzk9ky07+zN2+KIsp69flrec0Z1fbjuYX2w9kEpmTxG96OxVOe+0rvz0wd35h0f2pZLZ\n11m+7+KTc/4Z3XnoyYO54Y5teXLXQJJk06nL8xu/emKqqqqyr28k+/pGs+/QaPb1jWb3geGMjk/P\nPbYrT5h9LeLKE5pyaGgiew/O/h73DYzn2Y92VdXsPGhpqktrU12aGmoyMTmTtub6DIxMZHBkMkMj\nExkem0ptTVXqa2tSX1ed+rqa1NfWpGHu68Of66qzv3/2KPrA8ET6hycyMvbc05rraqqzenlz1na3\nZu3ylrnYPTQ0nh17h/KPT+xPkpy3sSvvu2RDmhtqMzI+ldHDHyPjU9mxbyi3P7A745PTaWmsza+f\nvy5XXnFmJkYn8tSegVx/25P55bbZC8Kcf3p3fvuik1NXW52h0ckMj05maHQyQ2OTGRufTn1ddRrr\na9PUUJPG+tkdAfW1NRmfnM7YxHTGJqbmvp6cmkljfU2aGmoPf8x+XV1VlX19o3OP8+6DI9l7cCRj\nE9NZ1lqfNctbnvnoakl7S3227x3Mk7sGsm33QJ7eM5iJqWeOrrQ112XD6vacsro9G9a0p3tZU3bv\nH86OfUPZsW8o2/cNpvfQ7ONcW1OVNctbcuKKtqxf0ZoTV7ams70xvYfnxe6DI9lzOIAPDU2kqiqp\nqa5OTU1Vag+fut1QV53lHU3pXtaU7mWNWdHZnO5ljWlvrs/gyGT6hyfSPzw+9386NjE9+zt1+Per\no6V+bvuwtWcgW3r6s6WnP7v2Dz/n/76hviYb17Tn1LUdOXVtR9Ysb8no+FSGx6Zm/1/GJjM8OpXJ\nqek0N9alpak2rY11aWma/airqc7gyEQGRydnPw9PZnB0IpNTM2ltqktbc33ajnxurstMpZKn9w7m\n6T2HP/YOZfTwGQlVSVZ1NefkVW05aVV7Tl7Vlomp6dxyX0/u37J/bkfIpeetyZqulvzsl3vy6OGd\nKSuWNeWdb16bt5zRnYmpmYyOTWV4bDIjY7Pzc2p6Js0NtWlurJ39dzTWprmhNvV1NZmeqWR6eibT\nM5VMHf48U6mkKlWprq6a2+FUVZXDf65KdZ71dVUyU5k98js+MT37+fDXM5VKGutr0lBXk4bDnxvr\nazMwPJHHdhzKY9v78viOQ3M7d46oranOKWvac8b6ZTnjxGWZnJrJbffvygNbZx+Hxvqa/OrZq9LR\nUp/bH9yVg4e3ISevass7Nq3J6q6W/OLJA3lw64HsOPyWNc/X1d6Yc0/tynkbu7KqqyWHBsfT9+yP\nofFMT8+ks60hnW0NOaGtMcvaGnJCW0Pq62oyNjGVsYnpw+uB2d/LI2Ob24HXUJum+ppUVVVlZHwq\nY+NTGZ2YXX50fCozM5XU1VanrrY69bU1qaurTn3t7DakUqnMrhsrSSWz25Aj/xfV1VWprqpKTXVV\nqqpn/w9qDn+vuvqZj9GxqfT2j2b/obG5z/v7RzM9U0l78zPbovaW2fk5OT0zt97Yd2g0vYfGMjE5\nPbcOeP4O0Y6W+kxOz2RyaiZT07PzZ/LweqOutjp1NdWpra1ObU3VXFBPT89kavqZ5Y9sb6urM/dv\nqq6umtsGTk7NZGJyJhNT05mYmsnE5HSqq6pm14t1sztJj8ytZHYeTkzOLjcxOZ3xw+Opf/bjfPjr\nmsNzeG4H6+FfxH19o3ly10Ce3NWfJ3cNzG1DjljWWp/1h9dvs0FWd3ibP3t0tlKZ3bHwzDbqme1U\nQ231c85SOqJSqWR8cjojY1Nz66Bn//4eedpfqeTwvKikvq4mLYd/n1ua6tLcWJuWxrq5s4uef/8z\nlcOP/5Hf+elKpmcqh9fBs+veI+vgZx7/2cd98vBjPzk1k6rqqtlt7OHHsr62OnV1NZmcms7Y+PRz\nto+jE9OpVCrPmbPV1VWpranOZW9Zn9Hh1/b2TQvhNQXYf/gP/yHveMc7cvnllydJPvShD+Uv/uIv\nsmHDhhdd/vUYYOOT0/n+3dvz5M5D2X1gZG5vVzK7Ivinl254w+6NeWjbwdx3+GIbb3/Tinm5wESl\nUsmhoYk8tXsg23YPZmh0Mk0NNXnzad05ZW37i+5dnalU8tjTh/KPT/RmarqS9pb6vO/ik4+5R/nh\nbQdzz+HxV1Ula5e35JS1HVnf3fKcFdP0TCX9Q7Mbw4MDz2wYxyefe3XHutrZI1vPf43X8z2+41Du\n+uXeVFUlnW0NWXt4A9K9rDGtTXVz82mmUsnU1MzchmUuopJnNoYzydizVpYj41MZHZvduB7ZUFZl\ndsV+xPjk7Mpp8kWuZNncUJsVnU1ZeUJTVnQ2p6O1Pn0D49nZO5Sd+2aPRBxRW1OVyzatecmL0BwY\nGMvN9+6ci6D62upsXNuR09Z3ZFnrM+c3D41O5omd/dmy89Dcssns43PeqV1Zv6L1Ob9nlUoluw+M\n5IEt++eecD9fdVXS2lyf1V3NWdfdmlUnNL3oBieZff3ivr7R9B4azeDo7JPd4bHJjI5N5fkruarM\nPmE+suGdedYTxiMb8unpmcy8xNqxuaE2Ha2zT86XtTSko60+yzuaUnOUOfvOTbNvzv6tW7fm8cNH\nso6ms60h77lgfS47b02aGmpfcAr1ngMjue/x3hdszEuorq5Ke3NdGutrM3g4YI+mKsmytoZ0dcz+\nXvQNjmf/odGXvE1DXU062xtyyur27D4wnJ29w3NPxI6mq70hXe2NmUkyPV1J//B4ZmYqmZmpZHJ6\n5jnzcT7U1lTNRl1nU1oaa7O/fyy9faPpH16c1yBUJVl5wmxwtTbVZce+oTy9d3BuR8mznbSqLe86\nf13edtaKuSOY9245kKd6DuWx7YeybffAC3ZWHS9aGmuz8oTmrDqhOfV11dl7cDR7+0bmourZutob\nc/r6jpy8un3uSeZMpZJdvcN5fMeh9PQOP2e9UV1VNbfzZ92KltTWVGfX/tn5uWv/sefoUlRbU5Wa\n6uoXbEefr7G+Jm3N9WlurM3MTCV9g+MZGp18ydssRfWHdwYt72jMTGX2cegbGJ97KQev3u9ctjG/\nc9FJiz2MF3ipADvmOU7P77NjnebxUj9sMf3B+1/6lJ83qg8u9gA4LvzrD2xa7CEsCd3dbblw07pX\nfLv3vk7XqywN5hdAWcc8uX7lypXZv3//3J/37duX5cuXL+igAAAAlqJjBtjFF1+cH/3oR0mShx9+\nOCtWrDjq678AAAA4umOegnj++efn7LPPzubNm1NVVZU///M/LzEuAACAJedlvQ8YAAAAr93Le4MV\nAAAAXjMBBgAAUMiSD7DPfe5zufLKK7N58+Y8+OCDiz0clqDHH388l19+ef7H//gfiz0UlqDPf/7z\nufLKK/OBD3wgP/7xjxd7OCwho6Oj+eM//uN8+MMfzgc/+MHccsstiz0klqCxsbG8613vyre//e3F\nHgpLyEMPPZTLLrssH/nIR/KRj3wkn/3sZxd7SK/IMS/CcTz7+c9/nqeffjrf/OY3s2XLlvzZn/1Z\nrrvuusUeFkvIyMhIPvvZz+bCCy9c7KGwBN1111154okn8s1vfjN9fX15//vfn/e85z2LPSyWiFtu\nuSXnnHNOfv/3fz89PT35V//qX+XXfu3XFntYLDFf+tKXsmyZ92Jlfo2MjOSKK67IVVddtdhDeVWW\ndIDdeeedufzyy5Mkp556agYGBjI0NOQy+syb+vr6fOUrX8lXvvKVxR4KS9AFF1yQc889N0nS0dGR\n0dHRTE9Pp6amZpFHxlLwm7/5m3Nf7969OytXrlzE0bAUbd26NVu2bMk73/nOxR4KS8zw8PBiD+E1\nWdKnIO7fvz+dnZ1zf+7q6kpvb+8ijoilpra2No2NjYs9DJaompqaNDc3J0muu+66XHbZZeKLebd5\n8+b86Z/+aT796U8v9lBYYq6++up86lOfWuxhsASNjIzk3nvvze/93u/lX/7Lf5m77rprsYf0iizp\nI2DPv8J+pVJJVVXVIo0G4NW56aab8q1vfStf//rXF3soLEF/+7d/m0ceeSSf/OQnc8MNN9hOMi++\n+93vZtOmTVm/fv1iD4Ul6Mwzz8zHP/7xvOtd78q2bdvyu7/7u/nxj3+c+vr6xR7ay7KkA2zlypXZ\nv3//3J/37duX5cuXL+KIAF6Z22+/Pddee22++tWvpq2tbbGHwxLy0EMPpaurK6tXr85ZZ52V6enp\nHDx4MF1dXYs9NJaAW2+9NTt27Mitt96aPXv2pL6+PqtWrcpFF1202ENjCdi4cWM2btyYJNmwYUOW\nL1+evXv3HjfBv6QD7OKLL84111yTzZs35+GHH86KFSu8/gs4bgwODubzn/98vvGNb3gRO/Punnvu\nSU9PT6666qrs378/IyMjzzltH16LL3zhC3NfX3PNNVm7dq34Yt5861vfysjISD760Y+mt7c3Bw4c\nOK5ex7qkA+z888/P2Wefnc2bN6eqqip//ud/vthDYol56KGHcvXVV6enpye1tbX50Y9+lGuuucaT\nZebFD37wg/T19eUTn/jE3PeuvvrqrFmzZhFHxVKxefPmXHXVVfnQhz6UsbGx/Mf/+B9TXb2kXxoO\nLBHvfve786d/+qf50Y9+lImJiXzmM585bk4/TJKqyvNfKAUAAMCCsKsLAACgEAEGAABQiAADAAAo\nRIABAAAUIsAAAAAKWdKXoQdg6brtttvy13/916murs7o6GjWrVuX//Sf/lM+/vGPp7+/Px0dHZmZ\nmUl9fX3+4i/+wuX7AXhdcBl6AI47ExMTufTSS/N3f/d3WbFiRZLkv/yX/5Lly5fn5ptvzh/+4R/O\nvenrNddck4GBgVx11VWLOWQASOIIGADHofHx8YyMjGR0dHTue5/85CeTJDfffPPc92ZmZrJnz56c\ndtppxccIAC/GETAAjkt//dd/nS996UvZtGlT3va2t+WKK67IKaecko985CNzpyDu27cvHR0d+epX\nv5r29vbFHjIACDAAjl99fX254447cvfdd+fGG2/Mv/23/zY33njjc05BvO222/KXf/mXuf766xd5\ntAAgwAA4To2OjqapqWnuz3fccUf+83/+z1m2bNlzAixJLrjggvz4xz9OZ2fnYgwVAOa4DD0Ax53b\nb789V155ZYaGhua+t3379px00kkvWPbRRx9NQ0OD+ALgdcERMACOS3/zN3+T733ve2lqakqlUklX\nV1euuuqq/Mmf/Mnca8CSZGpqKp/85Cdz/vnnL/KIAUCAAQAAFOMURAAAgEIEGAAAQCECDAAAoBAB\nBgAAUIgAAwAAKESAAQAAFCLAAAAAChFgAAAAhfz/I7ZJMgL0gEQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f91cb7f6d50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "collapsed = pd.read_table(get(1766))\n",
    "\n",
    "fig, ax = plt.subplots( nrows=1, ncols=1 ,figsize=(15, 10) ) \n",
    "sns.distplot(collapsed[\"SB\"]);\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that the Strand Bias is in majority <2 , we therefore applied a filter to the snpsift output to select the variants with a Strand Bias <2 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Import and cleaning of input files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The input files has been obtained by applying several filters on the variants annotation : \n",
    " * EFFECT!='intergenic_region' : Keep only the variants in intragenic regions\n",
    " * ',' not in ALT\n",
    " * strand_bias<2\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import files into the notebook : \n",
    " * The joined file is the file resulting of a collumn join on snpsift output collection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "joined = pd.read_table(get(2571))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output of SnpSift returns a tabulated files with header containing cracayers that may cause problems further into the analysis. We therefore need to replace the problematic caracters in the column names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for i in joined.columns.values : \n",
    "    joined=joined.rename(index=str, columns={i:i.replace('.','_').replace('[*]_','_')})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we may need the original version of the data later in the analysis we are creating a working copy of the Dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sequences=joined.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take a look at the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>SRR5363633_wtP_fq_CHROM</th>\n",
       "      <th>SRR5363633_wtP_fq_POS</th>\n",
       "      <th>SRR5363633_wtP_fq_REF</th>\n",
       "      <th>SRR5363633_wtP_fq_ALT</th>\n",
       "      <th>SRR5363633_wtP_fq_QUAL</th>\n",
       "      <th>SRR5363633_wtP_fq_SAF</th>\n",
       "      <th>SRR5363633_wtP_fq_SAR</th>\n",
       "      <th>SRR5363633_wtP_fq_SRF</th>\n",
       "      <th>SRR5363633_wtP_fq_SRR</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR5363715_cip_fq_SRP</th>\n",
       "      <th>SRR5363715_cip_fq_EPP</th>\n",
       "      <th>SRR5363715_cip_fq_DP</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_GENEID</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_GENE</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_EFFECT</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_AA</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_FUNCLASS</th>\n",
       "      <th>SRR5363715_cip_fq_SB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>.-1031233.0-GA</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>.-1053012.0-G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>.-1117713.0-G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>.-1135136.0-C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>.-1144646.0-C</td>\n",
       "      <td>NC_008463</td>\n",
       "      <td>1144646</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>8583.5</td>\n",
       "      <td>181</td>\n",
       "      <td>86</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.13803</td>\n",
       "      <td>17</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1661 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id SRR5363633_wtP_fq_CHROM SRR5363633_wtP_fq_POS  \\\n",
       "0  .-1031233.0-GA                       .                     .   \n",
       "1   .-1053012.0-G                       .                     .   \n",
       "2   .-1117713.0-G                       .                     .   \n",
       "3   .-1135136.0-C                       .                     .   \n",
       "4   .-1144646.0-C               NC_008463               1144646   \n",
       "\n",
       "  SRR5363633_wtP_fq_REF SRR5363633_wtP_fq_ALT SRR5363633_wtP_fq_QUAL  \\\n",
       "0                     .                     .                      .   \n",
       "1                     .                     .                      .   \n",
       "2                     .                     .                      .   \n",
       "3                     .                     .                      .   \n",
       "4                     C                     G                 8583.5   \n",
       "\n",
       "  SRR5363633_wtP_fq_SAF SRR5363633_wtP_fq_SAR SRR5363633_wtP_fq_SRF  \\\n",
       "0                     .                     .                     .   \n",
       "1                     .                     .                     .   \n",
       "2                     .                     .                     .   \n",
       "3                     .                     .                     .   \n",
       "4                   181                    86                     0   \n",
       "\n",
       "  SRR5363633_wtP_fq_SRR         ...          SRR5363715_cip_fq_SRP  \\\n",
       "0                     .         ...                              .   \n",
       "1                     .         ...                              .   \n",
       "2                     .         ...                              .   \n",
       "3                     .         ...                              .   \n",
       "4                     0         ...                            0.0   \n",
       "\n",
       "  SRR5363715_cip_fq_EPP SRR5363715_cip_fq_DP SRR5363715_cip_fq_EFF_GENEID  \\\n",
       "0                     .                    .                            .   \n",
       "1                     .                    .                            .   \n",
       "2                     .                    .                            .   \n",
       "3                     .                    .                            .   \n",
       "4               3.13803                   17                            .   \n",
       "\n",
       "  SRR5363715_cip_fq_EFF_GENE SRR5363715_cip_fq_EFF_EFFECT  \\\n",
       "0                          .                            .   \n",
       "1                          .                            .   \n",
       "2                          .                            .   \n",
       "3                          .                            .   \n",
       "4                          .                            .   \n",
       "\n",
       "  SRR5363715_cip_fq_EFF_AA SRR5363715_cip_fq_EFF_IMPACT  \\\n",
       "0                        .                            .   \n",
       "1                        .                            .   \n",
       "2                        .                            .   \n",
       "3                        .                            .   \n",
       "4                        .                            .   \n",
       "\n",
       "  SRR5363715_cip_fq_EFF_FUNCLASS SRR5363715_cip_fq_SB  \n",
       "0                              .                    .  \n",
       "1                              .                    .  \n",
       "2                              .                    .  \n",
       "3                              .                    .  \n",
       "4                              .                  0.0  \n",
       "\n",
       "[5 rows x 1661 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see that the dataframe has an big number of collumns, including the information in 15 columns for each variants. Most of the frames are empty, meaning the sample does not present the variant.\n",
    "Several information are redundant of won't be used in this analysis. In order to avoid carrying along data we won't need, we are creating lists of the data we won't use to filter out the dataFrame : \n",
    "Columns of interest :\n",
    " * A list of columns containing the sequence of reference\n",
    " * A list of collumn contining the gene impacted by the variation\n",
    " * A list of collumn contining the alternative sequence\n",
    " * A list containing the collumns we will need for this analysis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "regexREF=re.compile(\".*(REF).*\") # List of columns containing the reference allele of the variant\n",
    "listREF=list(filter(regexREF.match,joined.columns.values))\n",
    "listsamples= [ i.replace('_fq_REF','') for i in listREF ] #list of sample names\n",
    "regexGENE=re.compile(\".*(GENE).*\") # List of columns containing the gene impacted by the variant\n",
    "listGENE=list(filter(regexGENE.match,joined.columns.values))\n",
    "regexALT=re.compile(\".*(ALT).*\") # List of columns containing the alternative allele of the variant\n",
    "listALT=list(filter(regexALT.match,joined.columns.values))\n",
    "regexclean=re.compile(\".*(_GENE)|(IMPACT)|(FUNCLASS).*\") # List of columns we want to filter out (Neither gene impacted, impact nor functionnal class)\n",
    "listrefsalts=list(filter(lambda i: not regexclean.search(i),sequences.columns.values))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we identified our interesting collumn, we can delete the other ones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for i in listrefsalts[1:]:\n",
    "    del sequences[i]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>SRR5363633_wtP_fq_EFF_GENEID</th>\n",
       "      <th>SRR5363633_wtP_fq_EFF_GENE</th>\n",
       "      <th>SRR5363633_wtP_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363633_wtP_fq_EFF_FUNCLASS</th>\n",
       "      <th>SRR5363634_con_fq_EFF_GENEID</th>\n",
       "      <th>SRR5363634_con_fq_EFF_GENE</th>\n",
       "      <th>SRR5363634_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363634_con_fq_EFF_FUNCLASS</th>\n",
       "      <th>SRR5363635_con_fq_EFF_GENEID</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR5363713_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363713_cip_fq_EFF_FUNCLASS</th>\n",
       "      <th>SRR5363714_cip_fq_EFF_GENEID</th>\n",
       "      <th>SRR5363714_cip_fq_EFF_GENE</th>\n",
       "      <th>SRR5363714_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363714_cip_fq_EFF_FUNCLASS</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_GENEID</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_GENE</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_FUNCLASS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>.-1031233.0-GA</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>.-1053012.0-G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>.-1117713.0-G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>.-1135136.0-C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>.-1144646.0-C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 333 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id SRR5363633_wtP_fq_EFF_GENEID SRR5363633_wtP_fq_EFF_GENE  \\\n",
       "0  .-1031233.0-GA                            .                          .   \n",
       "1   .-1053012.0-G                            .                          .   \n",
       "2   .-1117713.0-G                            .                          .   \n",
       "3   .-1135136.0-C                            .                          .   \n",
       "4   .-1144646.0-C                            .                          .   \n",
       "\n",
       "  SRR5363633_wtP_fq_EFF_IMPACT SRR5363633_wtP_fq_EFF_FUNCLASS  \\\n",
       "0                            .                              .   \n",
       "1                            .                              .   \n",
       "2                            .                              .   \n",
       "3                            .                              .   \n",
       "4                            .                              .   \n",
       "\n",
       "  SRR5363634_con_fq_EFF_GENEID SRR5363634_con_fq_EFF_GENE  \\\n",
       "0                            .                          .   \n",
       "1                            .                          .   \n",
       "2                            .                          .   \n",
       "3                            .                          .   \n",
       "4                            .                          .   \n",
       "\n",
       "  SRR5363634_con_fq_EFF_IMPACT SRR5363634_con_fq_EFF_FUNCLASS  \\\n",
       "0                            .                              .   \n",
       "1                            .                              .   \n",
       "2                            .                              .   \n",
       "3                            .                              .   \n",
       "4                            .                              .   \n",
       "\n",
       "  SRR5363635_con_fq_EFF_GENEID              ...                \\\n",
       "0                            .              ...                 \n",
       "1                            .              ...                 \n",
       "2                            .              ...                 \n",
       "3                            .              ...                 \n",
       "4                            .              ...                 \n",
       "\n",
       "  SRR5363713_cip_fq_EFF_IMPACT SRR5363713_cip_fq_EFF_FUNCLASS  \\\n",
       "0                            .                              .   \n",
       "1                            .                              .   \n",
       "2                            .                              .   \n",
       "3                            .                              .   \n",
       "4                            .                              .   \n",
       "\n",
       "  SRR5363714_cip_fq_EFF_GENEID SRR5363714_cip_fq_EFF_GENE  \\\n",
       "0                            .                          .   \n",
       "1                            .                          .   \n",
       "2                            .                          .   \n",
       "3                            .                          .   \n",
       "4                            .                          .   \n",
       "\n",
       "  SRR5363714_cip_fq_EFF_IMPACT SRR5363714_cip_fq_EFF_FUNCLASS  \\\n",
       "0                            .                              .   \n",
       "1                            .                              .   \n",
       "2                            .                              .   \n",
       "3                            .                              .   \n",
       "4                            .                              .   \n",
       "\n",
       "  SRR5363715_cip_fq_EFF_GENEID SRR5363715_cip_fq_EFF_GENE  \\\n",
       "0                            .                          .   \n",
       "1                            .                          .   \n",
       "2                            .                          .   \n",
       "3                            .                          .   \n",
       "4                            .                          .   \n",
       "\n",
       "  SRR5363715_cip_fq_EFF_IMPACT SRR5363715_cip_fq_EFF_FUNCLASS  \n",
       "0                            .                              .  \n",
       "1                            .                              .  \n",
       "2                            .                              .  \n",
       "3                            .                              .  \n",
       "4                            .                              .  \n",
       "\n",
       "[5 rows x 333 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now reduced the number of collumns to 333."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this analysis we are interested in clustering the samples depending on the genes that are modified and how much they are modified.\n",
    "In order to do that we want to groupe the variants per genes and provide a modification score.\n",
    "\n",
    "\n",
    "The first Step is to create an additionnal column to the dataframe to store the information of the impacted gene, and then parse the dataframe to find the gene impacted by each variant.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "listsamples= [ i.replace('_fq_REF','') for i in listREF ]\n",
    "sequences['GENE']=\"NaN\"\n",
    "for i, row in sequences.iterrows() :\n",
    "    for j in listGENE :\n",
    "        if row[j]!='.':\n",
    "            sequences.loc[i,('GENE')]=row[j]\n",
    "            break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have regrouped the gene information in a single collumn we can delete the collumns containing the gene impacted for each sample, now redundant with the 'GENE' column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for i in listGENE:\n",
    "    del sequences[i]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the gene impacted for each variant, we need to quantify the impact of each variant on the gene. in order to do that we attributed a number reflecting the strength of each type of impact.\n",
    "\n",
    "In the Dataframe, we are therefore replacing the \"IMPACT\" column with numerical values depending on the strength of the effect : \n",
    " * No variant results on a value of 0 impact\n",
    " * A Low impact variant with a \"SILENT\" Functional class results on a value of 0 impact\n",
    " * A Low impact variant results on a value of 100 impact\n",
    " * A Medium impact variant results on a value of 200 impact\n",
    " * A High impact variant results on a value of 300 impact\n",
    " * A Modifier variant results on a value of 50 impact"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From http://snpeff.sourceforge.net/SnpEff_manual.html : \n",
    "    \n",
    "| Impact | Meaning | Example |\n",
    "| ------------ | ----------- | -----------: |\n",
    "| HIGH | The variant is assumed to have high (disruptive) impact in the protein, probably causing protein truncation, loss of function or triggering nonsense mediated decay. | stop_gained, frameshift_variant |\n",
    "| MODERATE      |   A non-disruptive variant that might change protein effectiveness.    | missense_variant, inframe_deletion |\n",
    "| LOW |     Assumed to be mostly harmless or unlikely to change protein behavior.      | synonymous_variant |\n",
    "| MODIFIER     |  Usually non-coding variants or variants affecting non-coding genes, where predictions are difficult or there is no evidence of impact.  | exon_variant, downstream_gene_variant |\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for i, row in sequences.iterrows() :\n",
    "    for j in listsamples :\n",
    "        if row[j+'_fq_EFF_IMPACT']=='.':\n",
    "            sequences.loc[i,(j+'_fq_EFF_IMPACT')]=0\n",
    "        elif 'LOW'in  row[j+'_fq_EFF_IMPACT'] and 'SILENT'in row[j+'_fq_EFF_FUNCLASS'] :\n",
    "            sequences.loc[i,(j+'_fq_EFF_IMPACT')]=0\n",
    "        elif 'LOW'in  row[j+'_fq_EFF_IMPACT'] :\n",
    "            sequences.loc[i,(j+'_fq_EFF_IMPACT')]=100\n",
    "        elif 'MODERATE'in  row[j+'_fq_EFF_IMPACT'] :\n",
    "            sequences.loc[i,(j+'_fq_EFF_IMPACT')]=200\n",
    "        elif 'HIGH'in  row[j+'_fq_EFF_IMPACT'] :\n",
    "            sequences.loc[i,(j+'_fq_EFF_IMPACT')]=300\n",
    "        else : \n",
    "            sequences.loc[i,(j+'_fq_EFF_IMPACT')]=50\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Now that we are done using the functionnal class information, we can remove the the funclass columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for j in listsamples :\n",
    "    del sequences[j+'_fq_EFF_FUNCLASS']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see what our data look like now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>SRR5363633_wtP_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363634_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363635_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363636_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363637_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363638_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363639_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363640_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363641_con_fq_EFF_IMPACT</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR5363707_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363708_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363709_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363710_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363711_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363712_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363713_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363714_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_IMPACT</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>.-1031233.0-GA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>.-1053012.0-G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>.-1117713.0-G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>.-1135136.0-C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>.-1144646.0-C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id SRR5363633_wtP_fq_EFF_IMPACT SRR5363634_con_fq_EFF_IMPACT  \\\n",
       "0  .-1031233.0-GA                            0                            0   \n",
       "1   .-1053012.0-G                            0                            0   \n",
       "2   .-1117713.0-G                            0                            0   \n",
       "3   .-1135136.0-C                            0                            0   \n",
       "4   .-1144646.0-C                            0                            0   \n",
       "\n",
       "  SRR5363635_con_fq_EFF_IMPACT SRR5363636_con_fq_EFF_IMPACT  \\\n",
       "0                            0                            0   \n",
       "1                            0                            0   \n",
       "2                            0                            0   \n",
       "3                            0                            0   \n",
       "4                            0                            0   \n",
       "\n",
       "  SRR5363637_con_fq_EFF_IMPACT SRR5363638_con_fq_EFF_IMPACT  \\\n",
       "0                            0                            0   \n",
       "1                            0                            0   \n",
       "2                            0                            0   \n",
       "3                            0                            0   \n",
       "4                            0                            0   \n",
       "\n",
       "  SRR5363639_con_fq_EFF_IMPACT SRR5363640_con_fq_EFF_IMPACT  \\\n",
       "0                            0                            0   \n",
       "1                            0                            0   \n",
       "2                            0                            0   \n",
       "3                            0                            0   \n",
       "4                            0                            0   \n",
       "\n",
       "  SRR5363641_con_fq_EFF_IMPACT ...  SRR5363707_cip_fq_EFF_IMPACT  \\\n",
       "0                            0 ...                             0   \n",
       "1                            0 ...                             0   \n",
       "2                            0 ...                             0   \n",
       "3                            0 ...                             0   \n",
       "4                            0 ...                             0   \n",
       "\n",
       "  SRR5363708_cip_fq_EFF_IMPACT SRR5363709_cip_fq_EFF_IMPACT  \\\n",
       "0                            0                            0   \n",
       "1                            0                            0   \n",
       "2                            0                            0   \n",
       "3                            0                            0   \n",
       "4                            0                            0   \n",
       "\n",
       "  SRR5363710_cip_fq_EFF_IMPACT SRR5363711_cip_fq_EFF_IMPACT  \\\n",
       "0                            0                            0   \n",
       "1                            0                            0   \n",
       "2                            0                            0   \n",
       "3                            0                            0   \n",
       "4                            0                            0   \n",
       "\n",
       "  SRR5363712_cip_fq_EFF_IMPACT SRR5363713_cip_fq_EFF_IMPACT  \\\n",
       "0                            0                            0   \n",
       "1                            0                            0   \n",
       "2                            0                            0   \n",
       "3                            0                            0   \n",
       "4                            0                            0   \n",
       "\n",
       "  SRR5363714_cip_fq_EFF_IMPACT SRR5363715_cip_fq_EFF_IMPACT GENE  \n",
       "0                            0                            0  NaN  \n",
       "1                            0                            0  NaN  \n",
       "2                            0                            0  NaN  \n",
       "3                            0                            0  NaN  \n",
       "4                            0                            0  NaN  \n",
       "\n",
       "[5 rows x 85 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have a dataframe containing the position of the variant, the gene impacted, and for each sample the strengh of the impact.\n",
    "\n",
    "We now whant to create a Matrix grouping the variants by gene impacted : the different variant impacts are added for a result reflecting the global impact of all variants on one gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Creation of an empty matri with header = list of samples\n",
    "matgene=pd.DataFrame(columns=sequences.columns.values[1:-1])\n",
    "# Group rows by gene, summing the row content for collumns contining variants\n",
    "matgene = sequences.groupby(['GENE'])[sequences.columns.values[1:-1]].sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take a look at our data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRR5363633_wtP_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363634_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363635_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363636_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363637_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363638_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363639_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363640_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363641_con_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363642_con_fq_EFF_IMPACT</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR5363706_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363707_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363708_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363709_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363710_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363711_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363712_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363713_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363714_cip_fq_EFF_IMPACT</th>\n",
       "      <th>SRR5363715_cip_fq_EFF_IMPACT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_00050</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_00490</th>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_00510</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PA14_00740</th>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>...</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 83 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            SRR5363633_wtP_fq_EFF_IMPACT  SRR5363634_con_fq_EFF_IMPACT  \\\n",
       "GENE                                                                     \n",
       "NaN                                    0                             0   \n",
       "PA14_00050                             0                             0   \n",
       "PA14_00490                           100                           100   \n",
       "PA14_00510                             0                             0   \n",
       "PA14_00740                           200                           200   \n",
       "\n",
       "            SRR5363635_con_fq_EFF_IMPACT  SRR5363636_con_fq_EFF_IMPACT  \\\n",
       "GENE                                                                     \n",
       "NaN                                    0                             0   \n",
       "PA14_00050                             0                             0   \n",
       "PA14_00490                           100                           100   \n",
       "PA14_00510                             0                             0   \n",
       "PA14_00740                           200                           200   \n",
       "\n",
       "            SRR5363637_con_fq_EFF_IMPACT  SRR5363638_con_fq_EFF_IMPACT  \\\n",
       "GENE                                                                     \n",
       "NaN                                    0                             0   \n",
       "PA14_00050                             0                             0   \n",
       "PA14_00490                           100                           100   \n",
       "PA14_00510                             0                             0   \n",
       "PA14_00740                           200                           200   \n",
       "\n",
       "            SRR5363639_con_fq_EFF_IMPACT  SRR5363640_con_fq_EFF_IMPACT  \\\n",
       "GENE                                                                     \n",
       "NaN                                    0                             0   \n",
       "PA14_00050                             0                             0   \n",
       "PA14_00490                           100                           100   \n",
       "PA14_00510                             0                             0   \n",
       "PA14_00740                           200                           200   \n",
       "\n",
       "            SRR5363641_con_fq_EFF_IMPACT  SRR5363642_con_fq_EFF_IMPACT  \\\n",
       "GENE                                                                     \n",
       "NaN                                    0                             0   \n",
       "PA14_00050                             0                             0   \n",
       "PA14_00490                           100                           100   \n",
       "PA14_00510                             0                             0   \n",
       "PA14_00740                           200                           200   \n",
       "\n",
       "                        ...               SRR5363706_cip_fq_EFF_IMPACT  \\\n",
       "GENE                    ...                                              \n",
       "NaN                     ...                                          0   \n",
       "PA14_00050              ...                                          0   \n",
       "PA14_00490              ...                                        100   \n",
       "PA14_00510              ...                                          0   \n",
       "PA14_00740              ...                                        200   \n",
       "\n",
       "            SRR5363707_cip_fq_EFF_IMPACT  SRR5363708_cip_fq_EFF_IMPACT  \\\n",
       "GENE                                                                     \n",
       "NaN                                    0                             0   \n",
       "PA14_00050                             0                             0   \n",
       "PA14_00490                           100                           100   \n",
       "PA14_00510                             0                             0   \n",
       "PA14_00740                           200                           200   \n",
       "\n",
       "            SRR5363709_cip_fq_EFF_IMPACT  SRR5363710_cip_fq_EFF_IMPACT  \\\n",
       "GENE                                                                     \n",
       "NaN                                    0                             0   \n",
       "PA14_00050                             0                             0   \n",
       "PA14_00490                           100                           100   \n",
       "PA14_00510                             0                             0   \n",
       "PA14_00740                           200                           200   \n",
       "\n",
       "            SRR5363711_cip_fq_EFF_IMPACT  SRR5363712_cip_fq_EFF_IMPACT  \\\n",
       "GENE                                                                     \n",
       "NaN                                    0                             0   \n",
       "PA14_00050                             0                             0   \n",
       "PA14_00490                           100                           100   \n",
       "PA14_00510                             0                             0   \n",
       "PA14_00740                           200                           200   \n",
       "\n",
       "            SRR5363713_cip_fq_EFF_IMPACT  SRR5363714_cip_fq_EFF_IMPACT  \\\n",
       "GENE                                                                     \n",
       "NaN                                    0                             0   \n",
       "PA14_00050                           200                             0   \n",
       "PA14_00490                           100                           100   \n",
       "PA14_00510                             0                             0   \n",
       "PA14_00740                           200                           200   \n",
       "\n",
       "            SRR5363715_cip_fq_EFF_IMPACT  \n",
       "GENE                                      \n",
       "NaN                                    0  \n",
       "PA14_00050                           200  \n",
       "PA14_00490                           100  \n",
       "PA14_00510                             0  \n",
       "PA14_00740                           200  \n",
       "\n",
       "[5 rows x 83 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matgene.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now have a Dataframe containing only the amount of modification of each gene in each sample. We are going to use these data to build a distance matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#distance =  lambda column1, column2: pd.np.linalg.norm(column1 - column2) # Euclidian Distance\n",
    "#distance = lambda column1, column2: scipy.stats.pearsonr(column1, column2)[0] # Pearson correlation coefficient\n",
    "distance = lambda column1, column2: (column1 - column2).abs().sum() / len(column1) # % Difference\n",
    "distdf = matgene.apply(lambda col1: matgene.apply(lambda col2: distance(col1, col2)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function that we want to use to builde a hierarchical tree doesn't take dataframes as imput. We therefore change the type and format of the distance matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dstmattril=np.tril(distdf) # make it a lower triangular matrix\n",
    "namessample= [str(x).replace(\"_fq_EFF_IMPACT\",\"\") for x in distdf.columns.values] #Remove unnecessary parts of sample names for more readability\n",
    "# make it a list of lists \n",
    "dstmattril= [list(x) for x in dstmattril] \n",
    "altDistmat=list()\n",
    "for i in range(0,len(distdf.columns.values)):\n",
    "    altDistmat.append(dstmattril[i][:i+1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have our distance matrix at the right format, we are building a clustering tree from the distance matrix (Here using neighor joining algorithm) and save the tree in newick format in a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "distmattree = _DistanceMatrix(names=namessample, matrix=altDistmat)\n",
    "constructor = DistanceTreeConstructor()\n",
    "tree = constructor.nj(distmattree)\n",
    "Phylo.write(tree,'pa_14_resistance.nwk', 'newick')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "put('pa_14_resistance.nwk')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
